simulator/src/main.rs

```diff
@@ -0,0 +1,632 @@
+    let payload = Payload {
+        stream: "action_status".to_string(),
+        device_id: None,
+        sequence: 0,
+        timestamp: SystemTime::UNIX_EPOCH.elapsed().unwrap().as_millis() as u64,
+        payload: json!({
+            "action_id": event.action_id,
+            "state": event.status,
+            "progress": event.progress,
+            "errors": []
+        }),
+    };
+
+    info!("Sending action response {} {} {}", event.action_id, event.progress, event.status);
+
+    let payload = serde_json::to_string(&payload)?;
+    client.send(payload).await?;
+    info!("Successfully sent action response");
+
+    Ok(())
+}
+
+pub async fn process_events(
+    events: &mut DelayQueue<Event>,
+    client: &mut SplitSink<Framed<TcpStream, LinesCodec>, String>,
+) -> Result<(), Error> {
+    if let Some(e) = events.next().await {
+        match e.into_inner() {
+            Event::DataEvent(event) => {
+                process_data_event(&event, events, client).await?;
+            }
+
+            Event::ActionResponseEvent(event) => {
+                process_action_response_event(&event, client).await?;
+            }
+        }
+    }
+
+    Ok(())
+}
+
+pub fn generate_action_events(action: Action, events: &mut DelayQueue<Event>) {
+    let action_id = action.action_id;
+
+    info!("Generating action events for action: {action_id}");
+    let now = Instant::now();
+
+    // Action response, 10% completion per second
+    for i in 1..10 {
+        let duration = Duration::from_secs(i as u64);
+        events.insert(
+            Event::ActionResponseEvent(ActionResponseEvent {
+                action_id: action_id.clone(),
+                progress: i * 10,
+                status: String::from("in_progress"),
+                timestamp: now + duration,
+            }),
+            duration,
+        );
+    }
+
+    let duration = Duration::from_secs(10);
+    events.insert(
+        Event::ActionResponseEvent(ActionResponseEvent {
+            action_id: action_id.clone(),
+            progress: 100,
+            status: String::from("Completed"),
+            timestamp: now + duration,
+        }),
+        duration,
+    );
+}
+
+#[tokio::main(flavor = "current_thread")]
+async fn main() -> Result<(), Error> {
+    let commandline = init();
+
+    let addr = format!("localhost:{}", commandline.port);
+    let path = read_gps_path(&commandline.paths);
+    let device = new_device_data(path);
+
+    let stream = TcpStream::connect(addr).await?;
+    let (mut data_tx, mut actions_rx) = Framed::new(stream, LinesCodec::new()).split();
+
+    let mut events = DelayQueue::new();
+
+    generate_initial_events(&mut events, Instant::now(), &device);
+
+    loop {
+        select! {
+            line = actions_rx.next() => {
+                let line = line.ok_or(Error::StreamDone)??;
+                let action = serde_json::from_str(&line)?;
+                generate_action_events(action, &mut events);
```

Comment:
This task is blocking the select! loop. Data won't be generated/sent. Spawn this in a task